File: subsystems/trigger.py
"""Trigger subsystem -- small NEO-driven cam that flicks a game piece into the shooter.

Teaching-bot proof of concept. The cam motor runs one full revolution per
"fire" -- the limit switch defines the cam's rest ("home") position, and
firing means: leave home, then come back to home. Two beam-break sensors
report loading status further upstream (not directly tied to the cam).
"""
from __future__ import annotations

import rev
import wpilib
from commands2 import Command, Subsystem, cmd

from constants import TriggerConstants


class Trigger(Subsystem):
    def __init__(self) -> None:
        super().__init__()

        self._cam_motor = rev.SparkMax(TriggerConstants.CAM_MOTOR_ID, rev.SparkMax.MotorType.kBrushless)
        cam_config = rev.SparkMaxConfig()
        cam_config.inverted(TriggerConstants.CAM_MOTOR_INVERTED)
        cam_config.setIdleMode(rev.SparkBaseConfig.IdleMode.kBrake)
        cam_config.smartCurrentLimit(TriggerConstants.CAM_CURRENT_LIMIT)
        self._cam_motor.configure(cam_config, rev.ResetMode.kResetSafeParameters, rev.PersistMode.kPersistParameters)

        self._limit_switch = wpilib.DigitalInput(TriggerConstants.LIMIT_SWITCH_DIO_PORT)
        self._beam_break_1 = wpilib.DigitalInput(TriggerConstants.BEAM_BREAK_1_DIO_PORT)
        self._beam_break_2 = wpilib.DigitalInput(TriggerConstants.BEAM_BREAK_2_DIO_PORT)

    def is_at_home(self) -> bool:
        raw = self._limit_switch.get()
        return not raw if TriggerConstants.LIMIT_SWITCH_INVERTED else raw

    def has_ball_at_stage_1(self) -> bool:
        raw = self._beam_break_1.get()
        return not raw if TriggerConstants.BEAM_BREAK_1_INVERTED else raw

    def has_ball_at_stage_2(self) -> bool:
        raw = self._beam_break_2.get()
        return not raw if TriggerConstants.BEAM_BREAK_2_INVERTED else raw

    def run_cam(self) -> None:
        self._cam_motor.set(TriggerConstants.CAM_FIRE_SPEED)

    def stop_cam(self) -> None:
        self._cam_motor.set(0.0)

    def fire_command(self) -> Command:
        """Runs the cam motor for one full revolution: leaves the home
        (limit-switch) position, then stops as soon as it returns to home.
        Has a timeout in case the limit switch never re-triggers (a jam or a
        broken wire), so this command can never run the motor forever."""
        has_left_home = False

        def _init() -> None:
            nonlocal has_left_home
            has_left_home = False

        def _finished() -> bool:
            nonlocal has_left_home
            if not has_left_home:
                if not self.is_at_home():
                    has_left_home = True
                return False
            return self.is_at_home()

        return (
            cmd.sequence(
                cmd.runOnce(_init, self),
                cmd.run(self.run_cam, self).until(_finished),
            )
            .withTimeout(TriggerConstants.FIRE_TIMEOUT_SECONDS)
            .finallyDo(lambda interrupted: self.stop_cam())
        )

    def periodic(self) -> None:
        wpilib.SmartDashboard.putBoolean("Trigger/AtHome", self.is_at_home())
        wpilib.SmartDashboard.putBoolean("Trigger/BallStage1", self.has_ball_at_stage_1())
        wpilib.SmartDashboard.putBoolean("Trigger/BallStage2", self.has_ball_at_stage_2())

File: src/main/java/frc/robot/subsystems/Trigger.java
package frc.robot.subsystems;

import static frc.robot.Constants.TriggerConstants.*;

import com.revrobotics.PersistMode;
import com.revrobotics.ResetMode;
import com.revrobotics.spark.SparkLowLevel.MotorType;
import com.revrobotics.spark.SparkMax;
import com.revrobotics.spark.config.SparkBaseConfig.IdleMode;
import com.revrobotics.spark.config.SparkMaxConfig;

import edu.wpi.first.wpilibj.DigitalInput;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
import edu.wpi.first.wpilibj2.command.SubsystemBase;

/**
 * Trigger subsystem -- small NEO-driven cam that flicks a game piece into the shooter.
 *
 * <p>Teaching-bot proof of concept. The cam has exactly one sensor: a limit switch that
 * defines its rest ("home") position. This subsystem only exposes plain actions/queries
 * (run the cam motor, read the switch/beam breaks) -- the "run until it's fired one
 * full revolution" logic is real state-machine behavior, so it lives in its own Command
 * class, commands/FireCommand.java, rather than here.
 */
public class Trigger extends SubsystemBase {

    private final SparkMax camMotor = new SparkMax(CAM_MOTOR_ID, MotorType.kBrushless);

    // DigitalInput reads a single digital (on/off) signal from a roboRIO DIO port --
    // the same class WPILib uses for any simple switch or break-beam sensor, since
    // electrically they're the same thing (a circuit that's either open or closed).
    private final DigitalInput limitSwitch = new DigitalInput(LIMIT_SWITCH_DIO_PORT);
    private final DigitalInput beamBreak1 = new DigitalInput(BEAM_BREAK_1_DIO_PORT);
    private final DigitalInput beamBreak2 = new DigitalInput(BEAM_BREAK_2_DIO_PORT);

    public Trigger() {
        SparkMaxConfig camConfig = new SparkMaxConfig();
        camConfig.inverted(CAM_MOTOR_INVERTED);
        // Brake mode (not DriveTrain's Coast): when the cam motor is commanded to 0, we
        // want it to stop and hold position immediately, not coast -- an idle cam
        // swinging freely could drift off "home" and throw off the next fire cycle's
        // home-switch reading.
        camConfig.idleMode(IdleMode.kBrake);
        camConfig.smartCurrentLimit(CAM_CURRENT_LIMIT);
        camMotor.configure(camConfig, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters);
    }

    /**
     * Every switch/beam-break getter here follows the same shape: read the raw
     * electrical signal, then flip it if that particular sensor's wiring reports
     * {@code true} for the opposite of what we mean (see the {@code *_INVERTED}
     * constants and their TODOs -- this is exactly the kind of thing that must be
     * checked on the real robot, since guessing wrong here silently inverts the
     * sensor's meaning). {@code cond ? a : b} is Java's ternary operator -- "if cond is
     * true, this whole expression's value is a, otherwise it's b" -- the closest Java
     * equivalent to Python's {@code a if cond else b}.
     */
    public boolean isAtHome() {
        boolean raw = limitSwitch.get();
        return LIMIT_SWITCH_INVERTED ? !raw : raw;
    }

    public boolean hasBallAtStage1() {
        boolean raw = beamBreak1.get();
        return BEAM_BREAK_1_INVERTED ? !raw : raw;
    }

    public boolean hasBallAtStage2() {
        boolean raw = beamBreak2.get();
        return BEAM_BREAK_2_INVERTED ? !raw : raw;
    }

    public void runCam() {
        camMotor.set(CAM_FIRE_SPEED);
    }

    public void stopCam() {
        camMotor.set(0.0);
    }

    @Override
    public void periodic() {
        SmartDashboard.putBoolean("Trigger/AtHome", isAtHome());
        SmartDashboard.putBoolean("Trigger/BallStage1", hasBallAtStage1());
        SmartDashboard.putBoolean("Trigger/BallStage2", hasBallAtStage2());
    }
}
